packages/engine/lib/error/src/report.rs

```diff
@@ -0,0 +1,293 @@
+    ///
+    /// [`I::Type`]: provider::TypeTag::Type
+    pub fn request<'p, I: 'static>(&'p self) -> RequestStack<'p, I>
+    where
+        I: TypeTag<'p>,
+    {
+        RequestStack::new(self)
+    }
+
+    /// Creates an iterator over the [`Frame`] stack requesting a reference of type `T`.
+    pub const fn request_ref<T: ?Sized + 'static>(&self) -> RequestStack<'_, Ref<T>> {
+        RequestStack::new(self)
+    }
+
+    /// Creates an iterator over the [`Frame`] stack requesting a value of type `T`.
+    pub const fn request_value<T: 'static>(&self) -> RequestStack<'_, Value<T>> {
+        RequestStack::new(self)
+    }
+}
+
+impl<S> fmt::Display for Report<S> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        let mut chain = self.frames();
+        let error = chain.next().expect("No error occurred");
+        fmt::Display::fmt(&error, fmt)?;
+        if let Some(cause) = chain.next() {
+            if fmt.alternate() {
+                write!(fmt, ": {cause}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<S> fmt::Debug for Report<S> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        if fmt.alternate() {
+            let mut debug = fmt.debug_struct("Report");
+            debug.field("frames", &self.frames());
+            #[cfg(feature = "backtrace")]
+            debug.field("backtrace", &self.backtrace());
+            debug.finish()
+        } else {
+            let mut chain = self.frames();
+            let error = chain.next().expect("No error occurred");
+            write!(fmt, "{error}")?;
+            write!(fmt, "\n             at {}", error.location())?;
+
+            for (idx, frame) in chain.enumerate() {
+                if idx == 0 {
+                    fmt.write_str("\n\nCaused by:")?;
+                }
+                write!(fmt, "\n   {idx}: {frame}")?;
+                write!(fmt, "\n             at {}", frame.location())?;
+            }
+
+            #[cfg(feature = "backtrace")]
+            if let Some(backtrace) = self.backtrace() {
+                write!(fmt, "\n\nStack backtrace:\n{backtrace}")?;
+            }
+
+            #[cfg(feature = "spantrace")]
+            if let Some(span_trace) = self.span_trace() {
+                write!(fmt, "\n\nSpan trace:\n{span_trace}")?;
+            }
+
+            Ok(())
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<E> From<E> for Report
+where
+    E: StdError + Send + Sync + 'static,
+{
+    #[track_caller]
+    fn from(error: E) -> Self {
+        #[cfg(feature = "backtrace")]
+        let backtrace = if error.backtrace().is_some() {
+            None
+        } else {
+            Some(Backtrace::capture())
+        };
+        Self::from_error(
+            Location::caller(),
+            #[cfg(feature = "backtrace")]
+            backtrace,
+            #[cfg(feature = "spantrace")]
+            Some(SpanTrace::capture()),
+            Error::Std(Box::new(error)),
+        )
+    }
+}
```

Comment:
Snippet from the `eyre` documentation:
> ### Wrapping Types That Don't impl `Error` (e.g. `&str` and `Box<dyn Error>`)
>
> Due to restrictions for coherence `Report` cannot impl `From` for types that don't impl `Error`. Attempts to do so will give "this type might implement Error in the future" as an error. As such, `wrap_err`, which uses `From` under the hood, cannot be used to wrap these types. Instead we encourage you to use the combinators provided for `Result` in `std`/`core`.
>
> For example, instead of this:
>
> ```rust,compile_fail
> use std::error::Error;
> use eyre::{WrapErr, Report};
>
> fn wrap_example(err: Result<(), Box<dyn Error + Send + Sync + 'static>>) -> Result<(), Report> {
>     err.wrap_err("saw a downstream error")
> }
> ```
>
> We encourage you to write this:
>
> ```rust
> use std::error::Error;
> use eyre::{WrapErr, Report, eyre};
>
> fn wrap_example(err: Result<(), Box<dyn Error + Send + Sync + 'static>>) -> Result<(), Report> {
>     err.map_err(|e| eyre!(e)).wrap_err("saw a downstream error")
> }
> ```

What do you think of this?